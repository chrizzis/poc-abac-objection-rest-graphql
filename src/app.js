import express from 'express'
import path from 'path'
import cors from 'cors'
import attachCurrentUser from './http-middleware/attachCurrentUser.js'
import { apiUrl } from '../config/index.js'
import userRoute from './routes/user.route.js'
import projectRoute from './routes/project.route.js'
import authRoute from './routes/auth.route.js'
const __dirname = path.resolve()

const app = express()

app.use(cors())
app.use(express.urlencoded({ extended: false }))
app.use(express.json())
app.use(express.static(path.join(__dirname, `/public`)))

app.use(authRoute)
app.use(apiUrl, attachCurrentUser)
app.use(apiUrl, userRoute)
app.use(apiUrl, projectRoute)

export default app