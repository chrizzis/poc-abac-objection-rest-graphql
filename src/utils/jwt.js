import jwt from 'jsonwebtoken'
// import dotenv from 'dotenv'
const { PUBLIC_KEY, PRIVATE_KEY } = process.env

const iss = 'jwt-node' // server?
const sub = 'jwt-node' // user.id
const aud = 'jwt-node' // client?

const defaultOptions = {
  issuer: iss,
  subject: sub,
  audience: aud,
  // expiresIn: '24h',
  expiresIn: '30s',
  algorithm: 'RS256',
}

// function expression vs declaration (function generateJwt...) - protect the class name from being reassigned within the class
// but is not hoisted, so any code that calls a function before it is expressed(?) will throw a ReferenceError
// payload = user: { id, username, roles } OR (refreshToken) { exp, sub }
const generateJWT = (payload, quick = false) => {
  let options
  if (payload) {
    if (payload.exp) { // refresh token
      const { expiresIn, subject, ...filteredOptions } = defaultOptions
      options = filteredOptions
    } else { // user token
      const { subject, ...filteredOptions } = defaultOptions
      console.debug(`generateJwt payload should be user.json: ${JSON.stringify(payload)}`)
      console.debug(`payload.id: ${payload.id.toString()}`)
      options = {
        ...filteredOptions,
        subject: payload.id.toString(),
        // subject: payload.id,
      }
      if (quick) {
        options.expiresIn = '10s'
      }
    }
  }
  return jwt.sign(payload, PRIVATE_KEY, options)
}

// TODO: keep track of request IP address origins to help detect token theft
// TODO: currently this is being used to verify user jwts, not refresh tokens
const verifyJWT = (payload) => {
  const { subject, ...filteredOptions } = defaultOptions
  return jwt.verify(payload, PUBLIC_KEY, filteredOptions)
}

export {
  generateJWT,
  verifyJWT,
}