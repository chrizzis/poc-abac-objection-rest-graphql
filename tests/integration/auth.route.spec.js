import request from 'supertest'
import app from '../../src/app.js'
import { currentEnv } from '../../config/index.js'
import pool from '../../src/db.js'
import { Model } from 'objection'
import { verifyJWT } from '../../src/utils/jwt.js'

// import Knex from 'knex'
// import knexConfigs from '../../knexfile.js'

// const knexConfig = knexConfigs[currentEnv]
// const knex = Knex(knexConfig);

describe("auth.route.js", () => {
  let knex
  beforeAll(async () => {
    // knex = Knex(knexConfig);
    knex = pool
    // TODO: knex.seed doesnt reset increment counter
    // knex.raw('ALTER TABLE ' + 'your_table' + ' AUTO_INCREMENT = 1');
    // Model.knex(knex) // not needed for tests to pass, not using objection here...
    // await knex.migrate.rollback();
    // await knex.migrate.latest(); // done in global setup
    // await knex.raw('TRUNCATE TABLE "User" RESTART IDENTITY CASCADE;')
    // await knex.raw('TRUNCATE TABLE "Project" RESTART IDENTITY CASCADE;')
    // await knex.seed.run(); // TODO: if kept here, can't destroy data in any tests that other tests rely on
  });
  beforeEach(async () => {
    // await knex.seed.run();
  });
  // TODO: this isnt needed b/c globalTeardown
  // afterAll(async () => {
  //   await knex.destroy()
  // })
  describe(`/login`, () => {
    test("successful login should return status 200 with token", async () => {
      const dbUsers = await knex('User');
      const dbuser = dbUsers[0]
      const { username, password } = dbuser
      const query = {
        username,
        password
      }

      const response = await request(app).post(`/login`).send(query).expect(200);
      console.debug(`response.body: ${JSON.stringify(response.body)}`)
      const { user, token } = response.body

      const decodedToken = verifyJWT(token)
      const { createdAt: edtca, exp, iat, ...resDecodedToken } = decodedToken
      expect(resDecodedToken).toMatchSnapshot()

      const { createdAt, ...resUser } = user
      // TODO: apparently the pg driver converts the db timestamp string to a js Date object
      // https://stackoverflow.com/questions/51082774/knex-silently-converts-postgres-timestamps-with-timezone-and-returns-incorrect-t
      expect(resUser).toMatchSnapshot()

    });
    test("login attempt with invalid credentials should return status 401", async () => {
      const query = {
        username: 'incorrect',
        password: 'incorrect',
      }
      await request(app).post(`/login`).send(query).expect(401);
    });
    test("login attempt with incomplete credentials should return status 401", async () => {
      const dbUsers = await knex('User');
      const { username, password } = dbUsers[0]
      await request(app).post(`/login`).send({ username }).expect(401);
      await request(app).post(`/login`).send({ password }).expect(401);
      await request(app).post(`/login`).send({}).expect(401);
    });
  });
});