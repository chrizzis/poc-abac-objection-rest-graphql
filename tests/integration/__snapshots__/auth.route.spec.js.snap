// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`auth.route.js /login successful login should return status 200 with token 1`] = `
Object {
  "aud": "jwt-node",
  "email": "alice@email.com",
  "firstName": "Alice",
  "id": 1,
  "iss": "jwt-node",
  "lastName": "Smith",
  "password": "alice",
  "role": "user",
  "sub": "1",
  "username": "alice-immutable",
}
`;

exports[`auth.route.js /login successful login should return status 200 with token 2`] = `
Object {
  "email": "alice@email.com",
  "firstName": "Alice",
  "id": 1,
  "lastName": "Smith",
  "password": "alice",
  "role": "user",
  "username": "alice-immutable",
}
`;
